### PR TITLE
Design/float comparison

### PR DESCRIPTION
**Pull request by HaicoDorenbos**
This PR deals with float comparison problems in the Roboteam_utils repository. == and != checks have been replaced by an absolute difference check to deal with float rounding problems. NOTE this PR should only be MERGED and REVIEWED once Design/lines #75 is MERGED.

### CODE REVIEW

File: include/roboteam_utils/HalfLine.h [hunk 1]
```diff
@@ -6,31 +6,40 @@
...
  *
  * @author Haico Dorenbos
  * @since 2020-04-22
  */
 class HalfLine {
     private:
-        Vector2 start; // The location where the HalfLine starts
-        Vector2 direction; // This vector indicates in which direction the HalfLine is infinite long
+        Vector2 start; // The location where the HalfLine starts.
+        Vector2 goesThrough; // An arbitrary other vector (different than start) through which the HalfLine goes.
```
Comment (RobotJente) [inline include/roboteam_utils/HalfLine.h:17]: The wording feels awkward in my opinion, I think it was better before. From this documentation it is also not clear to me if you mean that this vector is with respect to the origin, i.e. OA where A is a point on the line, or AB, where A and B are both on the line. Direction vector is universally understood to mean the latter, and therefore a better alternative, in my opinion.

File: include/roboteam_utils/HalfLine.h [hunk 2]
```diff
@@ -6,31 +6,40 @@
...
 
     public:
         /**
-         * Construct a new HalfLine object
-         * @param start The location where the HalfLine starts
-         * @param direction This vector indicates in which direction the HalfLine is infinite long
+         * Construct a new HalfLine object by using 2 different points on that HalfLine. Make sure that the given points are different.
+         * @param start The location where the HalfLine starts.
+         * @param goesThrough An arbitrary other vector through the HalfLine goes (make sure that it is different than the start point).
```
Comment (RobotJente) [inline include/roboteam_utils/HalfLine.h:23]: Same as before, I think this was more clear before.

File: include/roboteam_utils/HalfLine.h [hunk 3]
```diff
@@ -6,31 +6,40 @@
...
          */
-        constexpr HalfLine(const Vector2 &start, const Vector2 &direction) noexcept : start{start}, direction{direction} {};
+        explicit HalfLine(const Vector2 &start, const Vector2 &goesThrough);
 
         /**
-         * Compute the intersection point between this HalfLine and the parameter Line.
+         * Compute the intersection point between this HalfLine and the given Line.
```
Comment (RobotJente) [inline include/roboteam_utils/HalfLine.h:28]:
```suggestion
         * Compute the intersection point between this HalfLine and the Line given as parameter.
```
To avoid the unfortunate multiple definitions of parametrization (i.e. f(t) = (sin(t), cos(t)). This is nitpicking, though.

File: include/roboteam_utils/Line.h [hunk 4]
```diff
@@ -1,101 +1,126 @@
...
+     *	| |   ) || (      | (   ) || |   ) |  | |      | |   | || |   ) || (
+     *	| (__/  )| (____/\| )   ( || (__/  )  | (____/\| (___) || (__/  )| (____/\
+     *	(______/ (_______/|/     \|(______/   (_______/(_______)(______/ (_______/
      *
-     * @return Vector2 Vector representation of the direction of this vector
+     * The functions below are dead. Remove this tag if you use any of the functions and make sure to remove this tag at other places as well that will become alive by using any of
+     *the function below. Do not read/document/redesign/analyse/test/optimize/etc. any of this code, because it is a waste of your time! This code was not removed or placed at
+     *another branch, because other software developers are very attached to this code and are afraid that this code might be used at some day (but I think it won't be used at all
+     *and should be removed).
      */
```
Comment (RobotJente) [inline include/roboteam_utils/Line.h:105]: No.

File: include/roboteam_utils/Line.h [hunk 5]
```diff
@@ -1,101 +1,126 @@
...
+     * @param line The other (infinite) Line.
+     * @return In case there is no intersection (i.e. the Lines are parallel and distinct), return std::nullopt. In case there is a single intersection, return that point (Lines
+     * are non-parallel). In case there are multiple (infinitely many) intersections, return the intersection point closest to the origin (0,0).
      */
-    [[nodiscard]] double length() const;
+    [[nodiscard]] std::optional<Vector2> intersect(const Line &line) const;
 
     /**
-     * @brief Gets the length of the vector representation of this line
+     * Determine whether a point lies on this Line. This function is NOT protected against double/float rounding issues.
```
Comment (RobotJente) [inline include/roboteam_utils/Line.h:62]: Why not?

File: include/roboteam_utils/Line.h [hunk 6]
```diff
@@ -1,101 +1,126 @@
...
+     * @brief Checks whether \ref line intersects `this`
      *
-     * @return true True if line is vertical, will be true if isPoint is true
-     * @return false False if line is not vertical
+     * @param line Line to check against
+     * @return true True if \ref line intersects `this`
+     * @return false False if \ref line does not intersect `this`
      */
-    [[nodiscard]] bool isVertical() const;
+    // [[nodiscard]] bool doesIntersect(const LineSegment &line) const;
```
Comment (RobotJente) [inline include/roboteam_utils/Line.h:123]: I strongly disagree with commenting these functions, they are member functions that make a lot of sense to have, at least, to me, so I would not consider this dead code. If you want to hate on static utility functions that are unused, that is another matter. But member functions like this are not dead code, as long as the function makes sense to have for this class

File: include/roboteam_utils/Line.h [hunk 7]
```diff
@@ -171,37 +177,7 @@ class Line {
...
-
-    /**
-     * Get the relative position of pointOnLine on the given infinite line, i.e. compute t such that line1Start + (line1End - line1Start) * t = pointOnLine
-     *
-     * @param line1Start An arbitrary point on the first line.
-     * @param line1End An arbitrary other point on the first line (make sure that it is different than line1Start).
-     * @param pointOnLine A point that is located on the given line.
-     * @return The value t such that line1Start + (line1End - line1Start) * t = pointOnLine
-     */
-    static float relativePosition(const Vector2 line1Start, const Vector2 line1End, const Vector2 pointOnLine);
```
Comment (RobotJente) [inline include/roboteam_utils/Line.h:204]: Why are these removed?

File: include/roboteam_utils/Rectangle.h [hunk 8]
```diff
@@ -89,27 +89,12 @@ class Rectangle {
...
     [[nodiscard]] std::vector<Vector2> intersects(const LineSegment &line) const;
-    /**
-     * @brief Checks whether a given line intersects with `this`.
-     * Not the most efficient and may return the same point twice if the line 'exactly' intersects with the corner.
-     * In general prefer using LineSegments over Lines if you need intersections often.
-     * @param line line segment to check for
-     * @return vector with all of the points where the line segment intersects the rectangle
-     */
-
-    [[nodiscard]] std::vector<Vector2> intersects(const Line &line) const;
```
Comment (RobotJente) [inline include/roboteam_utils/Rectangle.h:100]: Why is this removed?

File: include/roboteam_utils/Rectangle.h [hunk 9]
```diff
@@ -89,27 +89,12 @@ class Rectangle {
...
      * @param line line segment to check for
      * @return True if `line` intersects with the rectangle
      */
     [[nodiscard]] bool doesIntersect(const LineSegment &line) const;
-    /**
-     * @brief Checks whether a given line intersects with `this`
-     * @param line line to check for
-     * @return True if `line` intersects with the rectangle
-     */
-    [[nodiscard]] bool doesIntersect(const Line &line) const;
```
Comment (RobotJente) [inline include/roboteam_utils/Rectangle.h:112]: Why is this removed?

File: include/roboteam_utils/Rectangle.h [hunk 10]
```diff
@@ -126,9 +111,40 @@ class Rectangle {
...
+     *	| |   | ||  __)   |  ___  || |   | |  | |      | |   | || |   | ||  __)
+     *	| |   ) || (      | (   ) || |   ) |  | |      | |   | || |   ) || (
+     *	| (__/  )| (____/\| )   ( || (__/  )  | (____/\| (___) || (__/  )| (____/\
+     *	(______/ (_______/|/     \|(______/   (_______/(_______)(______/ (_______/
+     *
+     * The functions below are dead. Remove this tag if you use any of the functions and make sure to remove this tag at other places as well that will become alive by using any of the
+     * function below. Do not read/document/redesign/analyse/test/optimize/etc. any of this code, because it is a waste of your time! This code was not removed or placed at another
+     * branch, because other software developers are very attached to this code and are afraid that this code might be used at some day (but I think it won't be used at all and should
+     * be removed).
+     */
```
Comment (RobotJente) [inline include/roboteam_utils/Rectangle.h:128]: Please no. I really disagree with having big tags in the codebase like this. It results in unnecessary clutter. You can put a -not maintained- tag over it if you must, but this is excessive and feels more like a slight against your team members because they don't agree with you

**Review by RobotJente — CHANGES_REQUESTED** [review]
My biggest problems with this PR are the commented out member functions, removal of overloaded functions, and the deadcode tags. Though there are good changes here, in its current form I don't think this should be merged.

**Comment by HaicoDorenbos** [comment]
> My biggest problems with this PR are the commented out member functions, removal of overloaded functions, and the deadcode tags. Though there are good changes here, in its current form I don't think this should be merged.

All of these comments are related to Design/lines #75. Regarding the dead code tag, we all agreed on using that instead of just removing the dead code

**Comment by RobotJente** [comment]
We agreed on using *a* tag, not this tag. We also clearly have different ideas of dead code, as I explained in my comments

**Comment by HaicoDorenbos** [comment]
> We agreed on using _a_ tag, not this tag. We also clearly have different ideas of dead code, as I explained in my comments

But despite that, your comment should be placed at https://github.com/RoboTeamTwente/roboteam_utils/pull/79

File: include/roboteam_utils/HalfLine.h (hunk 1 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/HalfLine.h:17]: I guess Rolf does not agree with you. He said that direction is a confusing term, since it also means the difference between both points.

File: include/roboteam_utils/HalfLine.h (hunk 2 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/HalfLine.h:23]: Same as before

File: include/roboteam_utils/Rectangle.h (hunk 8 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Rectangle.h:100]: For this you need to compute the intersection between a Line and LineSegment. And since the Line and LineSegment class are totally redesigned, I removed the Line with LineSegment intersect methods and did not add it again, since I do not spend time adjusting dead code.

File: include/roboteam_utils/Rectangle.h (hunk 9 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Rectangle.h:112]: Same reason as above.

File: include/roboteam_utils/Line.h (hunk 7 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Line.h:204]: The last 2 methods are not removed, but their definitions have been changed. The doesIntersect methods have been removed, because the Line and LineSegment classes have totally been redesigned and I do not spend time redesigning dead methods.

File: include/roboteam_utils/Line.h (hunk 5 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Line.h:62]: Forgot to update the documentation. It is now fixed in the other PR (#79)

File: include/roboteam_utils/HalfLine.h (hunk 3 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/HalfLine.h:28]: Problem is no longer present in PR (#79)

**Comment by HaicoDorenbos** [comment]
All these fixes are included in Redo of Design/lines #75 #79, so therefore this PR is closed.
